cmd/app/createca.go

```diff
@@ -0,0 +1,153 @@
+	Run: func(cmd *cobra.Command, args []string) {
+		log.Logger.Info("binding to PKCS11 HSM")
+		p11Ctx, err := crypto11.ConfigureFromFile("config/crypto11.conf")
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+		defer p11Ctx.Close()
+
+		rootID := []byte(viper.GetString("hsm-caroot-id"))
+
+		// Check if CA already exists (or a cert within the provided ID)
+		findCA, err := p11Ctx.FindCertificate(rootID, nil, nil)
+		if err != nil {
+			panic(err)
+		}
+		if findCA != nil {
+			panic("certificate already exists with this ID")
+		}
+
+		// Find the existing Key Pair
+		// TODO: We could make the TAG customizable
+		log.Logger.Info("finding slot for private key \"FulcioCA\"")
+		privKey, err := p11Ctx.FindKeyPair(nil, []byte("FulcioCA"))
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+		pubKey := privKey.Public().(crypto.PublicKey).(*ecdsa.PublicKey)
+
+		// Generate a Random Serial Number
+		// TODO: We could make it so this could be passed in by the user
+		serialNumber, err := rand.Int(rand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+		if err != nil {
+			panic(err)
+		}
+		rootCA := &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				Organization:  []string{viper.GetString("org")},
+				Country:       []string{viper.GetString("country")},
+				Province:      []string{viper.GetString("province")},
+				Locality:      []string{viper.GetString("locality")},
+				StreetAddress: []string{viper.GetString("street-address")},
+				PostalCode:    []string{viper.GetString("postal-code")},
+			},
+			NotBefore:             time.Now(),
+			NotAfter:              time.Now().AddDate(10, 0, 0),
+			IsCA:                  true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			BasicConstraintsValid: true,
+		}
+
+		caBytes, err := x509.CreateCertificate(rand.Reader, rootCA, rootCA, pubKey, privKey)
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+
+		pemBytes := pem.EncodeToMemory(&pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: caBytes,
+		})
+
+		log.Logger.Info("Root CA:")
+		fmt.Println(string(pemBytes))
+
+		certParse, err := x509.ParseCertificate(caBytes)
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+
+		// Import the root CA into the HSM
+		// TODO: We could make the TAG customizable
+		err = p11Ctx.ImportCertificateWithLabel(rootID, []byte("FulcioCA"), certParse)
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+		log.Logger.Info("root CA created with PKCS11 ID: ", viper.GetString("hsm-caroot-id"))
+
+		// Save out the file in pem format for easy import to CTL chain
+		if viper.IsSet("out") {
+			certOut, err := os.Create(viper.GetString("out"))
+			if err != nil {
+				log.Logger.Fatal(err)
+			}
+			if err := pem.Encode(certOut, &pem.Block{ //nolint
+				Type:  "CERTIFICATE",
+				Bytes: caBytes},
+			); err != nil {
+				log.Logger.Fatal(err)
```

Comment:
`certOut.Close()` needs to go above this line or deal with a deferred call to `Close`